File: 5-Problem-A-Day/170. Two Sum III - Data structure design.java
class TwoSum {
    Map<Integer, Integer> map = new HashMap<>();


    /** Initialize your data structure here. */
    public TwoSum() {
    }

    /** Add the number to an internal data structure.. */
    public void add(int number) {
        map.put(number, map.getOrDefault(number, 0) + 1);
    }

    /** Find if there exists any pair of numbers which sum is equal to the value. */
    public boolean find(int value) {
        for (int key : map.keySet()) {
            int remain = value - key;
            if (remain == key && map.get(key) >= 2)
                return true;
            if (remain != key && map.containsKey(remain))
                return true;
        }
        return false;
    }
}

/**
 * Your TwoSum object will be instantiated and called as such: TwoSum obj = new TwoSum();
 * obj.add(number); boolean param_2 = obj.find(value);
 */

File: 2021/170. Two Sum III - Data structure design.cpp
class TwoSum {
public:
    unordered_map<int, int> nums;
    /** Initialize your data structure here. */
    TwoSum() {
        
    }
    
    /** Add the number to an internal data structure.. */
    void add(int number) {      
        nums[number]++;
    }
    
    /** Find if there exists any pair of numbers which sum is equal to the value. */
    bool find(int value) {
        for (auto& iter : nums) {
            if ((iter.first > 0 && value < 0 && value < iter.first + INT_MIN) || (iter.first < 0 && value > 0 && value > iter.first + INT_MAX)) continue;
            if (nums.find(value - iter.first) != nums.end() && (value - iter.first != iter.first || nums[iter.first] > 1)) {                
                return true;
            }        
        }
        return false;
    }
};

/**
 * Your TwoSum object will be instantiated and called as such:
 * TwoSum* obj = new TwoSum();
 * obj->add(number);
 * bool param_2 = obj->find(value);
 */